Better tagging

itayVolk opened the pull request:
When tagging a post, you can now select multiple topic tags; an AI or someone else tag; as well as any future non-solved tags that aren't assigned to a menu. It also shows a dedicated message if no tags are selected and applies the tags on timeout.

https://github.com/user-attachments/assets/d6b85c00-067d-4923-a454-c5065cdfc554

itayVolk committed:
Improved tagging process

Now asks for AI and someone else tags, applies tags on timeout, shows a dedicated message for no tags, and allows selecting multiple tags on each screen. Still need to handle attempts to apply >5 tags.

itayVolk committed:
Implemented buttons option when tagging

Tagging can now include both single select button menus and multi select dropdown menus.

itayVolk committed:
Implemented multi select buttons

Multi select is now done using buttons (maintaining current UI look), tags (other than solved) which aren't placed in a menu are now prompted in a new menu. If more than 5 tags are added, the first 5 are added and the summary message lists the tags that weren't separately.

itayVolk committed:
style changes

itayVolk committed:
Fixed multi select timeout

itayVolk committed:
Added emoji and color to skip button, now changes text to "Done"

Run1e commented:
It's quite obvious that this method has become quite unwieldy. My initial implementation was hasty and I don't think it's a good idea to build on top of it (I mean the `tagask` method is absolutely massive at this point)

I think the primary issue is the lack of a proper way to handle state, so I think factoring out the operations you need to do on the list of tags/selections is a good place to start. Maybe a separate class with methods like `get_question`, `get_options`, `select_option`, `deselect_option`, etc in such a way that this `questions` object becomes a list of instances of that class: https://github.com/Run1e/AceBot/blob/master/cogs/ahk/ahk.py#L690-L707 - and clicking "done" would essentially just move to the next instance and rebuild the embed/options (which can be returned by `get_embed` or whatever)

What is the benefit of the AI & Someone Else tags? Not sure what that's supposed to help, if I know our members well I think it'll just make people mad and refuse to help people who admit to using AI

itayVolk committed:
Merge branch 'Run1e:master' into better_tagging

itayVolk commented:
I can try to convert it into a class or two (might need one for multi and one for single).
I think that even with AI/someone else tags people will still provide some help.

itayVolk committed:
Converted tagask to class based

itayVolk committed:
Merge branch 'Run1e:master' into better_tagging

Run1e committed:
refactor: refactor ask logic

Run1e committed:
Merge branch 'better_tagging' of github.com:itayVolk/AceBot into better_tagging

Run1e committed:
fix: fix broken tag stuff

Run1e commented:
@itayVolk can you check if the latest commits on this branch still work as you'd expect?

itayVolk committed:
Improved format.bat

Added "@echo off" to the top of format.bat

itayVolk commented:
@Run1e  Yes, the only difference (which I assume is intentional) is the removal of the ability to tag with any tags that aren't assigned to a question.

Run1e commented:
Yes that was intentional, if it works I'll merge this soon-ish

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -725,39 +818,46 @@ def check(inter: disnake.MessageInteraction):
-        content = (
-            "Thanks for tagging your post!\nYou can change the tags at any time by using `/retag`\n"
-        )
+        added_tags_list = list(added_tags)
```
@itayVolk if you want/have interest, it's possible to come up with some kind of algorithm to prioritize some tags over others if more than 5 is chosen. Here I just naively cast to a list and take the first five, something more intelligent could be done in theory

itayVolk replied:
I'll look at it later

itayVolk commented:
I found an issue with skipping in a single select menu (fixed), and you reverted the behavior on timeout to deleting the message without applying the tags instead of acting as if the rest of the menus were skipped.

Run1e committed:
Merge branch 'master' into better_tagging

Run1e committed:
feat: break on timeout instead of returning

Run1e commented:
> and you reverted the behavior on timeout to deleting the message without applying the tags instead of acting as if the rest of the menus were skipped.

Made it add the selected ones instead

itayVolk committed:
Updated which tags are selected when more than 5

Now picks single select tags first, and then the multi select tags in the order they were selected.
Also fixes issue with single select skip.

Run1e committed:
refactor: minor cleanup

itayVolk commented on cogs/ahk/ahk.py:
```diff
@@ -842,7 +838,7 @@ async def tagask(self, thread: disnake.Thread, inter: disnake.AppCmdInter | None
                 break
 
-            if len(result) > 1:
+            if isinstance(question, MultiSelect):
```
I thought making it depend on length would be better for future proofing.

Run1e committed:
feat: change selects, minor cleanup
